### PR TITLE
firedancer: communicate stake weights

### PR DESCRIPTION
**Pull request by mmcgee-jump**
We need to know stake weights to do two things,

 (a) Perform QoS for QUIC connections
 (b) Determine priority for outgoing shred distribution

We don't yet have the ability to calculate stake weights, so for now we communicate them via a shared memory primitive from the Solana service.

### CODE REVIEW

File: core/src/staked_nodes_updater_service.rs [hunk 1]
```diff
@@ -19,21 +20,55 @@ pub struct StakedNodesUpdaterService {
...
+    staked_nodes_overrides: &HashMap<Pubkey, u64>,
+) {
+    let mut values = staked_nodes.iter().filter(|(pubkey, _)| !staked_nodes_overrides.contains_key(pubkey)).collect::<Vec<_>>();
+    values.extend(staked_nodes_overrides.iter());
+    values.sort_by(|a, b| b.cmp(a));
+
+    let mvcc_data = mvcc.begin_write();
+    mvcc_data[0..8].copy_from_slice(&(values.len() as u64).to_le_bytes());
+
+    let max_elements = (mvcc_data.len() - 8) / 40;
```
Comment (ptaffet-jump) [inline core/src/staked_nodes_updater_service.rs:46]: Some comments or constants about these numbers would be good here

File: core/src/staked_nodes_updater_service.rs [hunk 2]
```diff
@@ -19,21 +20,55 @@ pub struct StakedNodesUpdaterService {
...
 }
 
+fn firedancer_write_stake(
+    mvcc: &mut Mvcc,
+    staked_nodes: &HashMap<Pubkey, u64>,
+    staked_nodes_overrides: &HashMap<Pubkey, u64>,
+) {
+    let mut values = staked_nodes.iter().filter(|(pubkey, _)| !staked_nodes_overrides.contains_key(pubkey)).collect::<Vec<_>>();
+    values.extend(staked_nodes_overrides.iter());
+    values.sort_by(|a, b| b.cmp(a));
```
Comment (lidatong) [inline core/src/staked_nodes_updater_service.rs:33]: why sorted? it still gets deser'd into an `fd_map_dynamic` right?

File: core/src/staked_nodes_updater_service.rs [hunk 3]
```diff
@@ -19,21 +20,55 @@ pub struct StakedNodesUpdaterService {
     thread_hdl: JoinHandle<()>,
 }
 
+fn firedancer_write_stake(
+    mvcc: &mut Mvcc,
```
Comment (lidatong) [inline core/src/staked_nodes_updater_service.rs:24]: will assume taking an &mut here and the Mvcc rust interface respects rust's rules...

File: core/src/staked_nodes_updater_service.rs [hunk 4]
```diff
@@ -19,21 +20,55 @@ pub struct StakedNodesUpdaterService {
...
+    mvcc_data[0..8].copy_from_slice(&(values.len() as u64).to_le_bytes());
+
+    let max_elements = (mvcc_data.len() - 8) / 40;
+    if values.len() > max_elements {
+        warn!("staked_nodes len {} exceeds max_elements {}", values.len(), max_elements);
+    }
+    for (i, value) in values.iter().enumerate().take(max_elements) {
+        let offset = 8 + i * 40;
+        mvcc_data[offset..offset + 8].copy_from_slice(&value.1.to_le_bytes());
+        mvcc_data[offset + 8..offset + 40].copy_from_slice(&value.0.to_bytes());
```
Comment (lidatong) [inline core/src/staked_nodes_updater_service.rs:53]: i think `fd_stake_deser` reads the pubkey first, so just making a note to update there too

**Review by lidatong — APPROVED** [review]
see comments but approved

File: core/src/staked_nodes_updater_service.rs (hunk 2 above)
Comment (mmcgee-jump) [reply core/src/staked_nodes_updater_service.rs:33]: added a comment

File: core/src/staked_nodes_updater_service.rs (hunk 4 above)
Comment (mmcgee-jump) [reply core/src/staked_nodes_updater_service.rs:53]: ok!

**Review by ptaffet-jump — APPROVED** [review]

**Pull request merged**